### PR TITLE
Merge vanilla server changes from latest binary into NPC.

### LINKED ISSUE #105: Butterfly Statue spawns catchable butterflies
Referencing: NyxStudios/TShock#1311

And https://github.com/NyxStudios/TShock/issues/1311#issuecomment-254800714

### LINKED ISSUE #1311: Butterfly Statue spawns catchable butterflies
Tested in 4.3.19 and 4.3.20, butterfly statues spawn butterflies that can be caught with either bug net. According to the wiki, critters spawned by critter statues should disappear instead of being caught. I have not yet been able to test this against stock TerrariaServer or with other statues.

### PR DESCRIPTION
**Pull request by DogooFalchion**
I went through a decompiled version of NPC and looked for any missing merges.  There are a few missing NPC names, some mob balance changes, and some other random changes to logic.  This also includes logic that checks for the issues reported in and fixes #105, and fixes NyxStudios/TShock#1311

### CODE REVIEW

**Comment by Ijwu** [comment]
LGTM but...

@Patrikkk @ProfessorXZ  You're the most familiar with the recent decompiles. Please check this for me as well.

Thanks very much for the PR, btw!

File: Terraria/NPC.cs [hunk 1]
```diff
@@ -88506,7 +88594,16 @@ public static void CatchNPC(int i, int who = -1)
 			}
 			if (Main.npc[i].catchItem > 0)
 			{
-				int arg_61_0 = Main.npc[i].type;
+				if (Main.npc[i].SpawnedFromStatue)
+				{
+					Vector2 position = Main.npc[i].Center - new Vector2(20f);
+					//Utils.PoofOfSmoke is probably client side magic.
```
Comment (Simon311) [inline Terraria/NPC.cs:88600]: Did you verify that? (that it does not require a packet being sent by us, or say a projectile generated on the server)

File: Terraria/NPC.cs [hunk 2]
```diff
@@ -84978,85 +85046,105 @@ public void NPCLoot()
...
+							{
+								Main.player[num5].name,
+								" has defeated the ",
+								NPC.killCount[num3],
+								"th ",
+								Lang.npcName(l, false),
+								"!"
+							});
+						}
+						if (Main.netMode == 0)
```
Comment (ivanbiljan) [inline Terraria/NPC.cs:85098]: Main.netMode == 0 is singleplayer mode, thus shouldn't be in here. Don't take it as criticism though, I'd just like the code to stay as clean as possible.

**Comment by ivanbiljan** [comment]
@Ijwu It's good.

File: Terraria/NPC.cs (hunk 1 above)
Comment (Patrikkk-zz) [reply Terraria/NPC.cs:88600]: PoofOfSmoke is creating new dust object. Method is non-existent in our version of TSAPI.

**Comment by Patrikkk-zz** [comment]
lgtm mate 👍  thank you for the contribution :)  @DogooFalchion

File: Terraria/NPC.cs (hunk 1 above)
Comment (Ijwu) [reply Terraria/NPC.cs:88600]: Does that mean it's client specific?

Comment (ivanbiljan) [reply Terraria/NPC.cs:88600]: Yes, it does.

**Comment by Ijwu** [comment]
Alright, thanks for the input guys. I'm going to leave this PR to be merged by another NyxStudios member due to the fact that it hasn't been compiled and run by any of us. I'm just introducing another level of oversight.

File: Terraria/NPC.cs (hunk 2 above)
Comment (DogooFalchion) [reply Terraria/NPC.cs:85098]: If you look below that was already in the source, I did not add it. the only lines that changed are 85051-85061

**Comment by dark22** [comment]
Hello, testing in a clean server, these changes seem to have Tower NPCs disappear when there are no nearby players. I did some tests and the previous build and works correctly without this problem.

**Comment by DogooFalchion** [comment]
You say works correctly, but are you comparing this to a previously broken tshock build or vanilla terraria?  There are a few lines that were changed from the previous version that appear to be related to being near/interacting with certain npcs.  It might be the intended vanilla behavior.

**Pull request merged**